### ISSUE TITLE
inconsistent point cloud rendering results with pytorch3D and other splatting method

### ISSUE THREAD

**Issue by zl548** [facebookresearch/pytorch3d#811]
I tried to use Pytorch3D to render a point cloud unprojected by a depth map from reference image (i.e., SfM aligned mono-depth) and render it into target viewpoint, but I found that there are significant differences between the image rendered by Pytorch3D and by other forward splatting algorithm.

The simplest demo code to reproduce the difference:
[point_cloud_render.tar.gz](https://github.com/facebookresearch/pytorch3d/files/7022975/point_cloud_render.tar.gz)
In script demo_bug.py, "img_1" is reference image and "depth_align" is the depth of reference image, "img_2" is target image we want to render. function "render_forward_rasterize" uses Pytorch3D, and "render_forward_splat" use softmax splatting from 
https://github.com/hperrot/splatting, and other logic such as point unprojection is exactly the same.

Here are the results:
This is ground truth image from target viewpoint :
![img_2](https://user-images.githubusercontent.com/7671455/130265355-77b163ec-a025-4d13-a727-3b6ac1ed53e9.jpg)
This is rendered image by Pytorch3D point cloud rasterization:
![wrp_img_rasterize](https://user-images.githubusercontent.com/7671455/130265364-f05055ec-ce72-4b33-a7e8-4ae080946b59.jpg)
This is rendered by splatting method from https://github.com/hperrot/splatting
![wrp_img_splat](https://user-images.githubusercontent.com/7671455/130265369-a7979e3c-6cc9-4d88-8357-161df5fd12d6.jpg)

If you flip between GT and warped image, you can see the image from splatting is more aligned with GT image, which I assume should be correct one.

I am not sure if this is bug of my code or bug of Pytorch3D, so it will be very helpful if someone could look into that.

Thanks

**Comment by zl548** [comment]
Moreover, more weird thing is that if we set R to identity and translation to 0, meaning that warped image should be exactly same as original reference image, I found that in the provided code above, splatting method gave me exactly same image as original image, but Pytorch3D gives me slightly different rendering as original image. This confuses me a lot....

You can try by adding following simple code to override R and t in demo_bug.py I provided:
R = torch.eye(3).float().cuda()
t = torch.zeros(3).float().cuda()

This is reference GT image: 
![img_1](https://user-images.githubusercontent.com/7671455/130267596-101dc318-658b-45c9-892a-336405260d73.jpg)
This is rendering from Pytorch3D:
![wrp_img_rasterize](https://user-images.githubusercontent.com/7671455/130267634-31c7710b-cdfc-431d-b81c-a4ef28cb8c99.jpg)
This is warped image from softmax splatting code:
![wrp_img_splat](https://user-images.githubusercontent.com/7671455/130267687-995c7625-c5da-4a65-bbc0-fea2800a8a41.jpg)

I can see there is distortion happening in Pytorch3D rendering result, but I cannot tell why....

**Comment by gkioxari** [comment]
I can take a look at this. However, the thing that pops immediately is that the compositors in the two cases are completely different. In the case of PyTorch3D, you are using an alpha compositor and in the other case you are using softmax. So a difference in outcome is expected.

**Comment by zl548** [comment]
I agreed that the splatting mechanism is different, which can make results somehow different, but the alignment with respect to the GT target image is very different (>=5 pixels drift) in the first case, which I don't expect it to happen since I assume final projected point 2D location should be very similar to each other if R,t,K are the same

**Comment by gkioxari** [comment]
Hi @zl548! I looked into your issue! There is a few things that are happening in your code and I also have some suggestions for how to use PyTorch3D more correctly!

**Issue in your code**
* The most important issue in your original code snippet is that you pass `R, t` into the cameras but you are already transforming the point cloud with `R, t` explicitly in your code. This means that you are making the renderer transform `pts3D_2` again with `R, t` leading to a double transform with the extrinsics. Instead, you should just do (Note that I provide the full code below that gets to the result).
```python
cameras = PerspectiveCameras(
            focal_length=((fx, fy),),
            principal_point=((px, py),),
            in_ndc=False,
            image_size=((height, width),),
            device=device,
)
```
The result is the following image which is very close to the output with the softmax splat
![img2_py3d](https://user-images.githubusercontent.com/4369065/130319459-a70aba89-c1e0-489b-ba50-96b3f730ba80.png)

**Working with PyTorch3D**
The following remarks are tips when working with PyTorch3D.
* Coordinate system: Note that PyTorch3D's coordinate system (used by the renderers) is `+X left, +Y up`. In your code, you use apply the inverse `K` in order to unproject points, thus assuming that world is `+X right, +Y down`.  You choose to adjust this coordinate system discrepancy by feeding in `(-fx, -fy)` which I don't really like but most importantly will lead to bugs. I'd recommend converting your points before feeding them into the renderer to PyTorch3D's coordinate system. Below I provide a clean way to work with PyTorch3D by converting all data properly to obey the coordinate system conventions of the library. 

```python
device = torch.device("cuda:0")
ckpt = torch.load("/tmp/output/test_bug.tar")

depth_align = ckpt["depth_align"]
R = ckpt["R"]
t = ckpt["t"]
K = ckpt["K"]
Kinv = K.inverse()
src_ims = ckpt["img_1"]
dst_ims = ckpt["img_2"]

# M converts from your coordinate to PyTorch3D's coordinate system
M = torch.eye(3).to(R.device)
M[0, 0] = -1.0
M[1, 1] = -1.0
# convert R, t from your to Py3D coordinate system
R = M @ R @ M.inverse()
t = M @ t

fx, fy, px, py = K[0, 0], K[1, 1], K[0, 2], K[1, 2]
height, width = src_ims.shape[1:3]

x = np.arange(width)
y = np.arange(height)
coord = np.stack(np.meshgrid(x, y), -1)
coord = np.concatenate((coord, np.ones_like(coord)[:, :, [0]]), -1)  # z=1
coord = coord.astype(np.float32)
coord = torch.as_tensor(coord, dtype=torch.float32, device=device)
coord = coord[None]  # bs, h, w, 3

D = depth_align[:, :, :, None, None]
pts3D_1 = D * Kinv[None, None, None, ...] @ coord[:, :, :, :, None]
# pts3D_1 live in your coordinate system. Convert them to Py3D's
pts3D_1 = M[None, None, None, ...] @ pts3D_1
# from reference to targe tviewpoint
pts3D_2 = R[None, None, None, ...] @ pts3D_1 + t[None, None, None, :, None]
depth_2 = pts3D_2[:, :, :, 2, :]  # b,1,h,w

pts3D_2_flat = pts3D_2.view(pts3D_2.shape[0], -1, 3)
src_ims_flat = src_ims.view(src_ims.shape[0], -1, 3)

point_cloud = Pointclouds(points=pts3D_2_flat, features=src_ims_flat)

cameras = PerspectiveCameras(
    focal_length=((fx, fy),),
    principal_point=((px, py),),
    in_ndc=False,
    image_size=((height, width),),
    device=device,
)

raster_settings = PointsRasterizationSettings(
    image_size=(height, width),
    radius=1.0 / height * 2.0,
)

rasterizer = PointsRasterizer(cameras=cameras, raster_settings=raster_settings)

renderer = PointsRenderer(rasterizer=rasterizer, compositor=AlphaCompositor())

render_img = renderer(point_cloud)

rgb = render_img[0, :, :, :3]
Image.fromarray((rgb.cpu().numpy() * 255).astype(np.uint8)).save("/tmp/output/img2_py3d.png")
```
* Note that if you wanted to pass `R, t` to the cameras instead of manually transforming `pts3D_1` to `pts3D_2` in your code. Initially, you were doing both which lead to a double transform. In the code above, I keep your manual transform and remove `R, t` from the cameras. Alternatively, you can remove your manual transform and keep `R, t` in the cameras. In that case, you have to transpose `R` as we assume row order for the rotation matrices instead of column order. These details along with coordinate systems are always the nits to be mindful of when working with a 3D codebase and require tremendous care.

**Comment by zl548** [comment]
Thanks. This fix works for me!

**Comment by GNAQ** [comment]
This comprehensive and considerate response works for me as well. Thank you very much.
